doc/warm-reboot/NSF_Manager_HLD.md

```diff
@@ -0,0 +1,292 @@
+
+
+#### NSF Details Registration
+
+
+```
+Component = <name>
+Docker = <name>
+Freeze = <True>/<False>
+Checkpoint = <True>/<False>
+Reconciliation = <True>/<False>
+```
+
+
+Components that want to participate in warm-boot orchestration need to register the above details with NSF Manager. NSF Manager will use these registration details to determine the components that are going to participate in the orchestrated shutdown sequence and monitor reconciliation statuses during bootup. If a component doesn’t register with the NSF Manager then it will continue to operate normally until it is shutdown in [Phase 4](#phase-4-prepare-and-perform-reboot). Components that modify the switch state should register with NSF Manager because they can change the state of other components such as Orchagent, Syncd etc. that participate in the warm reboot orchestration and thus they can impact the warm reboot process.
+
+Components that want to participate in an orchestrated shutdown during warm reboot need to set _freeze = true_. NSF Manager will wait for the quiescence of all components that have _freeze = true_ in their registration. If _checkpoint = True_ only then will NSF Manager wait for the component to complete checkpointing. Components that want NSF Manager to monitor their reconciliation status need to set _reconciliation = true_.
+
+
+#### Shutdown Orchestration
+
+During warm reboot, NSF Manager will orchestrate switch shutdown in a multi-phased approach that ensures that the switch stack is in a stable state before the intent/state is saved. It will perform shutdown in the following phases:
+
+
+
+*   Phase 1: Freeze components & wait for switch quiescence
+*   Phase 2: Perform state verification (optional)
+*   Phase 3: Trigger checkpointing
+*   Phase 4: Prepare and perform reboot
+
+
+##### Phase 1: Freeze Components & Wait for Switch Quiescence
+
+
+
+![alt_text](img/freeze.png)
+
+NSF Manager will send freeze notification to all registered components and will wait for the quiescence of only those components that have set _freeze = true_. Upon receiving the freeze notification, the component will complete processing the current request queue and stop generating new intents. Stopping new intents from being generated means that boundary components should stop processing requests from external components (external events) and all components should stop their periodic timers that generate new requests (internal events). For example:
+
+
+
+*   UMF will stop listening to new gRPC requests from the controller.
+*   P4RT will stop listening to new gRPC requests from the controller and stop packet I/O.
```

Comment:
This design introduces the high-level principles for the new warm-boot orchestration framework. I wanted to limit the scope of this design to these core principles. That is why this section talks about the actions of different components at a very high-level, thereby providing a brief idea of stopping self-sourced events.

From P4RT perspective, the P4Runtime spec doesn't talk about reboots in general. During cold reboot, the client observes the same behavior as a server shutdown. Ideally, the client shouldn't observe different behavior during warm reboot. This is because from the P4Runtime lens, the server was shutdown (thus unavailable for sometime) and came back up in the same state as before thereby maintaining the read-write symmetry. I would be happy to discuss this further in the System Orchestration workgroup or in a separate thread.